api/end_device.proto

```diff
@@ -198,7 +198,7 @@ message ADRSettings {
     // Transmission power index to use.
     uint32 tx_power_index = 2 [(validate.rules).uint32.lte = 15];
     // Number of retransmissions.
-    uint32 nb_trans = 3 [(validate.rules).uint32.lte = 15];
+    uint32 nb_trans = 3 [(validate.rules).uint32 = {gte: 1, lte: 15}];
```

Comment:
Not necessarily. We validate fields only if they are in the field mask, and also we validate the full entity at the end.

So if you had `NbTrans` = 1 before, and you patch the data rate index, you won't get an error, since we validate only your data rate index, and the resulting object (with all fields) is also valid.

Reply:
No it does not. And I re-tested this just to make sure I'm not taking crazy pills:

```bash
adriansmares@Adrian-Stefans-MacBook-Pro-3 lorawan-stack % go run ./cmd/ttn-lw-cli dev set app1 eui-1111111111111111 --unset mac-settings.use-adr,mac-settings.adr-margin
{
  "ids": {
    "device_id": "eui-1111111111111111",
    "application_ids": {
      "application_id": "app1"
    },
    "dev_eui": "1111111111111111",
    "join_eui": "0000000000000000"
  },
  "created_at": "2022-08-18T16:26:18.117Z",
  "updated_at": "2022-08-18T16:27:07.484792Z",
  "network_server_address": "localhost",
  "mac_settings": {}
}

adriansmares@Adrian-Stefans-MacBook-Pro-3 lorawan-stack % go run ./cmd/ttn-lw-cli dev set app1 eui-1111111111111111 --mac-settings.adr.mode.static.data-rate-index 2 --mac-settings.adr.mode.static.nb-trans 1
{
  "ids": {
    "device_id": "eui-1111111111111111",
    "application_ids": {
      "application_id": "app1"
    },
    "dev_eui": "1111111111111111",
    "join_eui": "0000000000000000"
  },
  "created_at": "2022-08-18T16:26:18.117Z",
  "updated_at": "2022-08-18T16:27:38.994301Z",
  "network_server_address": "localhost",
  "mac_settings": {
    "adr": {
      "static": {
        "data_rate_index": 2,
        "nb_trans": 1
      }
    }
  }
}

adriansmares@Adrian-Stefans-MacBook-Pro-3 lorawan-stack % go run ./cmd/ttn-lw-cli dev set app1 eui-1111111111111111 --mac-settings.adr.mode.static.data-rate-index 3             
{
  "ids": {
    "device_id": "eui-1111111111111111",
    "application_ids": {
      "application_id": "app1"
    },
    "dev_eui": "1111111111111111",
    "join_eui": "0000000000000000"
  },
  "created_at": "2022-08-18T16:26:18.117Z",
  "updated_at": "2022-08-18T16:27:47.275406Z",
  "network_server_address": "localhost",
  "mac_settings": {
    "adr": {
      "static": {
        "data_rate_index": 3
      }
    }
  }
}

adriansmares@Adrian-Stefans-MacBook-Pro-3 lorawan-stack % go run ./cmd/ttn-lw-cli dev get app1 eui-1111111111111111 --mac-settings.adr                             
{
  "ids": {
    "device_id": "eui-1111111111111111",
    "application_ids": {
      "application_id": "app1"
    },
    "dev_eui": "1111111111111111",
    "join_eui": "0000000000000000"
  },
  "created_at": "2022-08-18T16:26:18.117Z",
  "updated_at": "2022-08-18T16:28:27.330455Z",
  "network_server_address": "localhost",
  "mac_settings": {
    "adr": {
      "static": {
        "data_rate_index": 3,
        "nb_trans": 1
      }
    }
  }
}

adriansmares@Adrian-Stefans-MacBook-Pro-3 lorawan-stack % go run ./cmd/ttn-lw-cli dev set app1 eui-1111111111111111 --unset mac-settings.adr                       
{
  "ids": {
    "device_id": "eui-1111111111111111",
    "application_ids": {
      "application_id": "app1"
    },
    "dev_eui": "1111111111111111",
    "join_eui": "0000000000000000"
  },
  "created_at": "2022-08-18T16:26:18.117Z",
  "updated_at": "2022-08-18T16:29:26.313721Z",
  "network_server_address": "localhost",
  "mac_settings": {}
}

adriansmares@Adrian-Stefans-MacBook-Pro-3 lorawan-stack % go run ./cmd/ttn-lw-cli dev set app1 eui-1111111111111111 --mac-settings.adr.mode.static.data-rate-index 3
error:pkg/errors:validation (invalid `mac_settings`: embedded message failed validation)
    reason=embedded message failed validation
    field=mac_settings
    name=EndDeviceValidationError
    correlation_id=929822419f0e4eab9062853901ab4364
--- error:pkg/errors:validation (invalid `adr`: embedded message failed validation)
    field=adr
    name=MACSettingsValidationError
    reason=embedded message failed validation
    correlation_id=0b0a9d22811949ae8b33ce7f912d60c0
--- error:pkg/errors:validation (invalid `static`: embedded message failed validation)
    reason=embedded message failed validation
    field=static
    name=ADRSettingsValidationError
    correlation_id=05adc2dc3eaf4cc693e2cf0d628e6061
--- error:pkg/errors:validation (invalid `nb_trans`: value must be inside range [1, 15])
    name=ADRSettings_StaticModeValidationError
    reason=value must be inside range [1, 15]
    field=nb_trans
    correlation_id=4aae60aebf7146cf894a8beae4f17943
exit status 255
```

Reply:
And the reason why this works is because `ttnpb.UpdateEndDeviceRequest.ValidateContext` provides the field mask to `ValidateFields`:
https://github.com/TheThingsNetwork/lorawan-stack/blob/474da537ec188a71e1916d060f398d782ee9f651/pkg/ttnpb/end_device_validators.go#L40-L49

Reply:
OK thanks for clarifying